components/Resize-node.coffee

```diff
@@ -22,48 +23,107 @@ exports.getComponent = ->
     datatype: 'integer'
     description: 'New height'
     required: false
-  c.outPorts.add 'canvas',
-    datatype: 'string'
+
+  c.outPorts.add 'out',
+    datatype: 'all'
+    description: 'Resized buffer'
+  c.outPorts.add 'factor',
+    datatype: 'number'
+    description: 'Original over resized dimensions factor'
+    required: false
+  c.outPorts.add 'original',
+    datatype: 'object'
+    description: 'Original dimension'
+    required: false
+  c.outPorts.add 'resized',
+    datatype: 'object'
+    description: 'Resized dimension'
+    required: false
+  c.outPorts.add 'metadata',
+    datatype: 'object'
+    description: 'Extracted metadata while resizing'
+    required: false
   c.outPorts.add 'error',
     datatype: 'object'
     required: false
 
   noflo.helpers.WirePattern c,
-    in: ['path']
+    in: ['in']
     params: ['width', 'height']
-    out: ['canvas']
+    out: ['out', 'factor', 'original', 'resized', 'metadata']
     async: true
     forwardGroups: true
   , (payload, groups, out, callback) ->
+    if (not Buffer.isBuffer payload) and (typeof payload isnt 'string')
+      return callback Error 'Input is not a valid buffer nor image path'
     width = c.params.width
     height = c.params.height
-    if not width? and not height?
-      width = 256
-    path = payload
     try
-      inputBuffer = sharp path
+      inputBuffer = sharp payload
       inputBuffer.metadata (err, metadata) ->
         if err
           return callback err
-        inputBuffer
-        .resize width, height
-        .withMetadata()
-        .withoutEnlargement()
-        .toFormat 'png'
-        .toBuffer (err, outputBuffer, info) ->
-          if err
-            return callback err
-          # Create image with buffer as src
-          image = new Canvas.Image
-          image.src = outputBuffer
-          # Create a host canvas and draw on it
-          canvas = new Canvas(info.width, info.height)
-          ctx = canvas.getContext '2d'
-          ctx.drawImage image, 0, 0, info.width, info.height
-          canvas.originalWidth = metadata.width
-          canvas.originalHeight = metadata.height
-          out.send canvas
-          do callback
+        # Default value when nothing is specified
+        if not width? and not height?
+          # Deal with narrow or wide images
+          if metadata.width > metadata.height
+            width = c.defaultDimension
+          else
+            height = c.defaultDimension
+        # Try to preserve the same format, if there's EXIF
+        if metadata.exif?
+          inputBuffer
+          .resize width, height
+          .withMetadata()
+          .withoutEnlargement()
+          .toBuffer (err, outputBuffer, info) ->
+            if err
```

Comment:
Yeah, and actually we don't need two branches because component outputs `metadata`, so we can resize the input as PNG while keeping EXIF information.